openedx/core/lib/partitions/partitions_service.py

```diff
@@ -3,7 +3,57 @@
 user partitions.  It uses the user_service key/value store provided by the LMS runtime to
 persist the assignments.
 """
-from abc import ABCMeta, abstractproperty
+from abc import ABCMeta, abstractmethod
+from django.utils.translation import ugettext as _
+import logging
+
+from openedx.core.lib.partitions.partitions import NoSuchUserPartitionError, UserPartition, UserPartitionError
+
+
+log = logging.getLogger(__name__)
+
+
+ENROLLMENT_TRACK_PARTITION_ID = 50
+
+
+def get_course_user_partitions(course):
+    """
+    A method that returns all `UserPartitions` associated with a course, as a List.
+    This will include the ones defined in course.user_partitions, but it may also
+    include dynamically included partitions (such as the `EnrollmentTrackUserPartition`).
+    """
+    return course.user_partitions + _get_dynamic_partitions(course)
+
+
+def _get_dynamic_partitions(course):
+    return [_create_enrollment_track_partition(course)]
+
+
+def _create_enrollment_track_partition(course):
+    try:
+        enrollment_track_scheme = UserPartition.get_scheme("enrollment_track")
+    except UserPartitionError:
+        log.warning("No 'enrollment_track' scheme registered, EnrollmentTrackUserPartition will not be created.")
+        return
+
+    used_ids = set(p.id for p in course.user_partitions)
+    if ENROLLMENT_TRACK_PARTITION_ID in used_ids:
```

Comment:
@nasthagiri the ID situation here is tricky. I can't just pick an unused ID anymore because this partition/ID is never persisted, and xblock's group_access WILL store the partition ID, so it is crucial that the ID always be the same whenever this dynamic partition is returned.

Luckily we should have 0-99 free as indices, so I think that dynamic partitions just have to declare their IDs as constants within that range (and if an XML-authored course happens to hit one of these IDs, then we have to skip adding those dynamic partitions). Of course, this would require much commenting, but for now I'm using "50".